File: leetcode_python/threeSumClosest.py
# Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

#     For example, given array S = {-1 2 1 -4}, and target = 1.

#     The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).

class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        res = sys.maxint
        for i in range(len(nums)-1):
            v = nums[i]
            left,right = i+1,len(nums)-1
            while left < right:
                s = v+nums[left]+nums[right]
                if s == target:
                    return target
                elif s < target:
                    if target - s < abs(target - res):
                        res = s
                    left += 1
                else:
                    if s-target < abs(res-target):
                        res = s
                    right -= 1
        return res
            

File: leetcode_python/threesum_closest.js
// Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

//     For example, given array S = {-1 2 1 -4}, and target = 1.

//     The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    if(!nums || nums.length < 3){
        return -1;
    }
    nums.sort(function(x,y){return x-y});
    let best = nums[0]+nums[1]+nums[2];
    for(let i = 0; i < nums.length; i++){
        let l = i+1;
        let r = nums.length-1;
        while(l < r){
            let sum = nums[l]+nums[i]+nums[r];
            if(Math.abs(sum-target) < Math.abs(best - target)) best = sum;
            if(sum < target) l++;
            else r--;
        }
    }
    return best;
};